cloud-controller-manager/do/loadbalancers.go

```diff
@@ -275,27 +306,91 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+func getCertificateIDFromLB(lb *godo.LoadBalancer) string {
+	var id string
+	for _, rule := range lb.ForwardingRules {
+		if rule.CertificateID != "" {
+			id = rule.CertificateID
+			// CCM does not currently support multiple certificates on the
+			// loadbalancers it manages so we only need to grab the first one we find
+			break
+		}
+	}
+	return id
+}
+
+func (l *loadBalancers) checkAndUpdateLBAndServiceCerts(ctx context.Context, service *v1.Service, lbCertID, serviceCertID string) error {
+	if lbCertID != "" && lbCertID != serviceCertID {
+		lbCert, _, err := l.resources.gclient.Certificates.Get(ctx, lbCertID)
+		if err != nil {
+			respErr, ok := err.(*godo.ErrorResponse)
+			if ok && respErr.Response.StatusCode == http.StatusNotFound {
+				return nil
+			}
+			return fmt.Errorf("failed to get DO certificate for lb: %s", err)
+		}
+
+		if lbCert.Type == certTypeLetsEncrypt {
+			ensureServiceAnnotation(service, annDOCertificateID, lbCertID)
+		}
+	}
+
+	return nil
+}
+
+func (l *loadBalancers) updateLoadBalancer(ctx context.Context, lb *godo.LoadBalancer, service *v1.Service, nodes []*v1.Node) (*godo.LoadBalancer, error) {
+	// call buildLoadBalancerRequest for its error checking; we have to call it
+	// again just before actually updating the loadbalancer in case
+	// checkAndUpdateLBAndServiceCerts modifies the service
+	_, err := l.buildLoadBalancerRequest(ctx, service, nodes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build load-balancer request: %s", err)
+	}
+
+	lbCertID := getCertificateIDFromLB(lb)
+	serviceCertID := getCertificateID(service)
+	err = l.checkAndUpdateLBAndServiceCerts(ctx, service, lbCertID, serviceCertID)
+	if err != nil {
+		return nil, err
+	}
+
 	lbRequest, err := l.buildLoadBalancerRequest(ctx, service, nodes)
 	if err != nil {
-		return fmt.Errorf("failed to build load-balancer request: %s", err)
+		return nil, fmt.Errorf("failed to build load-balancer request: %s", err)
 	}
 
-	lb, err := l.retrieveAndAnnotateLoadBalancer(ctx, service)
+	lbID := lb.ID
+	lb, _, err = l.resources.gclient.LoadBalancers.Update(ctx, lb.ID, lbRequest)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to update load-balancer with ID %s: %s", lbID, err)
 	}
 
-	_, _, err = l.resources.gclient.LoadBalancers.Update(ctx, lb.ID, lbRequest)
+	return lb, nil
+}
+
+// UpdateLoadBalancer updates the load balancer for service to balance across
+// the droplets in nodes.
+//
+// UpdateLoadBalancer will not modify service or nodes.
+func (l *loadBalancers) UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (err error) {
+	patcher := newServicePatcher(service)
+	defer func() {
+		if err == nil {
+			err = patcher.Patch(l.resources.kclient, service)
+			if err != nil {
+				err = fmt.Errorf("failed to patch service: %s", err)
+			}
+		}
+	}()
```

Comment:
Since we have implemented the deferred method three times, how about attaching it to `servicePatcher` so that we can do a direct `defer.Patch(...)` call (and move the error checking / handling into the method too)?